src/vcpkg/binarycaching.cpp

```diff
@@ -1255,6 +1255,159 @@ namespace
+                                 StringView package_name,
+                                 StringView package_version,
+                                 const Path& download_path,
+                                 MessageSink& sink) const
+        {
+            Command cmd = base_cmd(src, package_name, package_version, "download");
+            cmd.string_arg("--path").string_arg(download_path);
+            return run_az_artifacts_cmd(cmd, sink);
+        }
+
+        ExpectedL<Unit> publish(const AzureUpkgSource& src,
+                                StringView package_name,
+                                StringView package_version,
+                                const Path& package_dir,
+                                StringView description,
+                                MessageSink& sink) const
+        {
+            Command cmd = base_cmd(src, package_name, package_version, "publish");
+            cmd.string_arg("--description").string_arg(description);
+            cmd.string_arg("--path").string_arg(package_dir);
+            return run_az_artifacts_cmd(cmd, sink);
+        }
+
+        ExpectedL<Unit> run_az_artifacts_cmd(const Command& cmd, MessageSink& sink) const
+        {
+            RedirectedProcessLaunchSettings show_in_debug_settings;
+            show_in_debug_settings.echo_in_debug = EchoInDebug::Show;
+            return cmd_execute_and_capture_output(cmd, show_in_debug_settings)
+                .then([&](ExitCodeAndOutput&& res) -> ExpectedL<Unit> {
+                    if (res.exit_code == 0)
+                    {
+                        return {Unit{}};
+                    }
+                    if (res.output.find("you need to run the login command") != std::string::npos)
+                    {
+                        sink.println(Color::warning,
+                                     msgFailedVendorAuthentication,
+                                     msg::vendor = "Universal Packages",
+                                     msg::url = docs::troubleshoot_binary_cache_url);
+                    }
+                    return LocalizedString::from_raw(std::move(res).output);
+                });
+        }
+        Path az_cli;
+    };
+
+    struct AzureUpkgPutBinaryProvider : public IWriteBinaryProvider
+    {
+        AzureUpkgPutBinaryProvider(const ToolCache& cache, MessageSink& sink, std::vector<AzureUpkgSource>&& sources)
+            : m_azure_tool(cache, sink), m_sources(sources)
+        {
+        }
+
+        size_t push_success(const BinaryPackageWriteInfo& request, MessageSink& msg_sink) override
+        {
+            size_t count_stored = 0;
+
+            auto package_version = format_version_for_nugetref(request.version.text, request.package_abi);
+            std::string package_name = fmt::format("{}-{}", request.spec.name(), request.spec.triplet());
+            std::string package_description = "Cached package for " + package_name;
```

Comment:
Does this need to be localized?

Reply:
I don't think it needs to be. I could see other information going into the description such as build output.